### ISSUE TITLE
Forwarding Connection header breaks after updating grpc/grpc-go to 1.42.0

### ISSUE THREAD

**Issue by MakDon** [grpc-ecosystem/grpc-gateway#2447]
## 🐛 Bug Report

grpc-gateway breaks after updating `grpc/grpc-go` to 1.42.0

After the update introduced by [this PR](https://github.com/grpc/grpc-go/pull/4803),  it will report `stream terminated by RST_STREAM with error code: PROTOCOL_ERROR` if `connection` in headers, which breaks down all of the http connections that comes with `connect:keep-alive`

## To Reproduce

Step 1:
    Update `google.golang.org/grpc` to `v1.42.0` in go mod

Step 2：
    invoke http request with header `connect:keep-alive`

## Expected behavior

grpc-gateway works well or reports warning, instead of returning `PROTOCOL_ERROR`

## Actual Behavior

it returns `stream terminated by RST_STREAM with error code: PROTOCOL_ERROR`

## Your Environment

go.mod: 
`google.golang.org/grpc v1.42.0`, `github.com/grpc-ecosystem/grpc-gateway/v2 v2.7.0`

**Comment by johanbrandhorst** [comment]
Hm, this is curious, what should the behavior be here? Should the grpc-gateway strip these headers? I'm surprised we're forwarding this header, it should only forward explicitly added headers or permanent IANA headers (listed here: https://github.com/grpc-ecosystem/grpc-gateway/blob/094a6fe78b3ca888297d090185cdf30f0e42e157/runtime/context.go#L261). Would you be willing to submit a test and a fix for this?

**Comment by MakDon** [comment]
> Hm, this is curious, what should the behavior be here? Should the grpc-gateway strip these headers? I'm surprised we're forwarding this header, it should only forward explicitly added headers or permanent IANA headers (listed here:
> 
> https://github.com/grpc-ecosystem/grpc-gateway/blob/094a6fe78b3ca888297d090185cdf30f0e42e157/runtime/context.go#L261
> 
> ). Would you be willing to submit a test and a fix for this?

ok, I'd like to submit a PR for it later.

**Comment by MakDon** [comment]
> Hm, this is curious, what should the behavior be here? Should the grpc-gateway strip these headers? I'm surprised we're forwarding this header, it should only forward explicitly added headers or permanent IANA headers (listed here:
> 
> https://github.com/grpc-ecosystem/grpc-gateway/blob/094a6fe78b3ca888297d090185cdf30f0e42e157/runtime/context.go#L261
> 
> ). Would you be willing to submit a test and a fix for this?

I found that my server breaks because i was using `func(key string) (string, bool) {return key, true}` header matcher for backward compatibility, which passthrough `Connect` header to grpc server and deserves a `RST_STREAM`

`Connection` header does not belong to `permanent request headers maintained by IANA`, so it seems not to be a good idea to add it into `isPermanentHTTPHeader` function body. And it would not breaks servers that uses `DefaultHeaderMatcher` cause it is not begins with `Grpc-Metadata-`

But for a further concern, is it a good idea that provides a backward-Compatible header matcher, which drops "malformed" headers for grpc and returns others headers as-is? It can be implemented like this:

```go
func PassThroughHeaderMatcher(key string) (string, bool) {
	key = textproto.CanonicalMIMEHeaderKey(key)
	if isMalformedHTTPHeader(key) {
		return MetadataPrefix + key, true
	} else {
		return key, false
	}
}
```

Thanks very much.

**Comment by johanbrandhorst** [comment]
That is an interesting question. It would be nice to automatically help users who might suffer from this problem, but I'm also hesitant to do this implicitly. Maybe we could detect if the header matcher would match `Connection` on startup and log a warning?

**Comment by MakDon** [comment]
> That is an interesting question. It would be nice to automatically help users who might suffer from this problem, but I'm also hesitant to do this implicitly. Maybe we could detect if the header matcher would match `Connection` on startup and log a warning?

Agree. I tend to add a pre-matcher to match `Connection` and log a warning. Is it probable sulution?

**Comment by johanbrandhorst** [comment]
I think we can just try running any matcher with the `Connection` header when creating it in https://github.com/grpc-ecosystem/grpc-gateway/blob/master/runtime/mux.go#L115-L119 and log a warning if it matches. Would you be willing to contribute such a fix?

**Comment by MakDon** [comment]
Ok i am working on it.

**Comment by johanbrandhorst** [comment]
Fixed in #2455
